### PR TITLE
feat: make dot generic to support bf16/f16/f32 with one `dot_distance` interface. 

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

**Comment by eddyxu** [comment]
Auto-vectorization or (unrolling) kicks in for (f16/f32/f64) this time

```
// No auto-vectoration & unrolling
Dot(half::binary16::f16, arrow_artiy)
                        time:   [4.5868 s 4.6175 s 4.6470 s]
                        change: [-2.0101% -0.8736% +0.0291%] (p = 0.13 > 0.10)
                        No change in performance detected.

// Have auto-vectorization & unrolling
Dot(half::binary16::f16)
                        time:   [350.51 ms 351.10 ms 351.72 ms]
                        change: [-91.171% -91.141% -91.114%] (p = 0.00 < 0.10)
                        Performance has improved.
```

Surprisingly, the auto-vectorization is even faster than our manual SIMD

```
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high severe

Dot(f32)                time:   [82.276 ms 83.007 ms 84.023 ms]
                        change: [-89.987% -89.914% -89.834%] (p = 0.00 < 0.10)
                        Performance has improved.
Found 2 outliers among 10 measurements (20.00%)
  1 (10.00%) high mild
  1 (10.00%) high severe

Benchmarking Dot(f32, SIMD): Warming up for 3.0000 s
Warning: Unable to complete 10 samples in 5.0s. You may wish to increase target time to 6.1s or enable flat sampling.
Dot(f32, SIMD)          time:   [104.13 ms 108.07 ms 117.54 ms]
                        change: [-16.354% +5.8254% +34.378%] (p = 0.69 > 0.10)
                        No change in performance detected.
```

File: rust/.cargo/config.toml [hunk 1]
```diff
@@ -27,7 +27,7 @@ rustflags = [
 ]
 
 [target.x86_64-unknown-linux-gnu]
-rustflags = ["-C", "target-cpu=haswell", "-C", "target-feature=+avx2,+fma"]
+rustflags = ["-C", "target-cpu=haswell", "-C", "target-feature=+avx2,+fma,+f16c"]
 
 [target.aarch64-apple-darwin]
-rustflags = ["-C", "target-cpu=apple-m1", "-C", "target-feature=+neon"]
+rustflags = ["-C", "target-cpu=apple-m1", "-C", "target-feature=+neon,+fp16"]
```
Comment (eddyxu) [inline rust/.cargo/config.toml:33]: Verified M1 has `fp16`.

**Review by westonpace — APPROVED** [review]
Minor suggestions and a question about one todo, but otherwise looks good

File: rust/lance-linalg/src/simd/f32.rs [hunk 2]
```diff
@@ -565,11 +566,12 @@ impl SIMD<f32, 16> for f32x16 {
...
-            let mask = _mm512_cmpeq_ps_mask(self.0, tgt);
-            if mask != 0 {
-                return Some(mask.trailing_zeros() as i32);
-            }
+            // let tgt = _mm512_set1_ps(val);
+            // let mask = _mm512_cmpeq_ps_mask(self.0, tgt);
+            // if mask != 0 {
+            //     return Some(mask.trailing_zeros() as i32);
+            // }
+            todo!()
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/lance-linalg/src/simd/f32.rs:574]: We dont need it for now. Can not build Lance release with avx512 yet.

File: rust/lance-linalg/src/distance/dot.rs [hunk 3]
```diff
@@ -15,98 +15,162 @@
...
+                let x2 = f32x16::load_unaligned(x.as_ptr().add(16));
+                let x3 = f32x16::load_unaligned(x.as_ptr().add(32));
+                let x4 = f32x16::load_unaligned(x.as_ptr().add(48));
+
+                let y1 = f32x16::load_unaligned(y.as_ptr());
+                let y2 = f32x16::load_unaligned(y.as_ptr().add(16));
+                let y3 = f32x16::load_unaligned(y.as_ptr().add(32));
+                let y4 = f32x16::load_unaligned(y.as_ptr().add(48));
+
+                sum16 += (x1 * y1 + x2 * y2) + (x3 * y3 + x4 * y4);
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/lance-linalg/src/distance/dot.rs:150]: This is force the low 4 registers adding together, and high 4 registers adding together, to reduce data dependency / increase parallelism at register level.

File: rust/lance-linalg/benches/dot.rs [hunk 4]
```diff
@@ -37,8 +36,9 @@ fn dot_arrow_artiy<T: ArrowNumericType>(x: &PrimitiveArray<T>, y: &PrimitiveArra
 
 fn run_bench<T: ArrowNumericType>(c: &mut Criterion)
 where
-    T::Native: Real + FromPrimitive + Sum,
+    T::Native: FromPrimitive + FloatToArrayType,
     NativeAdapter<T>: From<T::Native>,
+    <T::Native as FloatToArrayType>::ArrowType: Dot,
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/lance-linalg/benches/dot.rs:41]: Let me double-check again.

[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/lance-linalg/benches/dot.rs:41]: This was suggested by the compiler, and if we do `T: Dot` , it turns out that it requires to make more changes of the other two constraints.

**Pull request merged**